api/contrib/envoy/extensions/filters/http/golang/v3alpha/golang.proto

```diff
@@ -0,0 +1,195 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.golang.v3alpha;
+
+import "google/protobuf/any.proto";
+import "google/protobuf/struct.proto";
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.golang.v3alpha";
+option java_outer_classname = "GolangProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/golang/v3alpha";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: golang extension filter]
+// golang extension filter.
+//
+// In the below example, we configured the go plugin 'auth' and 'limit' dynamic libraries into
+// Envoy, which can avoid rebuilding Envoy.
+//
+// * Develop go-plugin
+//
+// We can implement the interface of :ref:`StreamFilter <contrib.go_plugin_api.plugin.StreamFilter>`
+// API by the GO language to achieve the effects of Envoy native filter.
+//
+// The filter based on the `APIs implementation
+// <https://github.com/mosn/envoy-go-filter-samples/blob/master/simple/filter.go>`_ for details.
```

Comment:
Sorry what I mean is there should be some test filter in this repo, as well as a test that actually exercises that it works. Can we make that happen? Thank you.

/wait

Reply:
> Sorry what I mean is there should be some test filter in this repo, as well as a test that actually exercises that it works. Can we make that happen? Thank you.
> 

@mattklein123  Well, at the moment we put this working example in the external [repository](https://github.com/mosn/envoy-go-filter-samples/), because the implementation will continue to iterate, so it is easy to manage, and we can migrate it when it is stable. Here's a quick start doc that can be use go-plugins to affect request processing in Envoy : https://github.com/mosn/envoy-go-filter-samples/blob/master/README.md .

Reply:
> OK well we won't merge here until there is something stable so feel free to report back when it's ready. Thank you.

@mattklein123  So could you please review the API first? At present, it is only a specific implementation based on the API that is  continuously iterating, and the API itself has been basically completed.  The implementation we will be pushed after completes the API. Thank you.

Reply:
At a high level it looks OK but it needs a lot more comments, and a working example will be helpful. I will wait until there is something more fleshed out to look at.

Reply:
> At a high level it looks OK but it needs a lot more comments, and a working example will be helpful. I will wait until there is something more fleshed out to look at.

Hi @mattklein123 Here is a working example  https://github.com/mosn/envoy-go-filter-samples/blob/master/README.md, we can try it in the following way.

Build go plugin
======
```
git clone git@github.com:mosn/envoy-go-filter-samples.git  && cd envoy-go-filter-samples && make build
```

Run go plugin
======
```
Notice: only works on amd64 architecture CPU yet.

docker run -d -p 10000:10000 \
    -v `pwd`/libgolang.so:/usr/local/envoy-go-extension/libgolang.so \
    -v `pwd`/envoy.yaml:/etc/envoy/envoy-golang.yaml \
    mosnio/envoy-go-extension:latest
# 1. it will proxy to httpbin.org/header
# we will see the request header and response header setted by Go.

$ curl 'http://127.0.0.1:10000/headers' -v
*   Trying 127.0.0.1:10000...
* Connected to 127.0.0.1 (127.0.0.1) port 10000 (#0)
> GET /headers HTTP/1.1
> Host: 127.0.0.1:10000
> User-Agent: curl/7.79.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< date: Sun, 30 Oct 2022 02:02:16 GMT
< content-type: application/json
< content-length: 260
< server: envoy
< access-control-allow-origin: *
< access-control-allow-credentials: true
< x-envoy-upstream-service-time: 1014
< rsp-header-from-go: bar-test
<
{
  "headers": {
    "Accept": "*/*",
    "Host": "httpbin.org",
    "Req-Header-From-Go": "foo-test",
    "User-Agent": "curl/7.79.1",
    "X-Amzn-Trace-Id": "Root=1-635ddb28-7c27d0446b0281c10c3170d8",
    "X-Envoy-Expected-Rq-Timeout-Ms": "15000"
  }
}

# 2. local reply from Go
$ curl 'http://127.0.0.1:10000/localreply'
forbidden from go, path: /localreply
```

Reply:
The example needs to be self contained and in this repo.

Reply:
> The example needs to be self contained and in this repo.

@mattklein123 I've pushed the example to this repo in this [commit](https://github.com/envoyproxy/envoy/pull/22573/commits/eeda2e9b6ba11a16175f436e83230462fb6fc933).

Reply:
Now it can be exercised in CI and verify golang plugin examples succeeded. we can look at this [log](https://dev.azure.com/cncf/envoy/_build/results?buildId=120207&view=logs&j=f1a49366-51c0-540f-0294-f73c2bb311a9&t=7fd036c9-61ec-5b59-3645-e90950dc67a6&l=7479) for details.

Reply:
I want to see a completely self contained example in this repo, that is exercised in tests. Having the sandbox is also good.

Reply:
@mattklein123 I have switched the mode of compiling the example to `go vendor` mode and avoid pulling external code during CI tests. And also migrated all the code of the example to the directory `examples/golang/envoy-go-filter-samples/`. Thank you.

Reply:
1) Please figure out how to avoid vendoring proto code, etc.
2) Remove all references to MOSN, and simplify the example to something as small as possible.

Thanks

/wait

Reply:
@mattklein123   I've already committed the API and implementation to this PR. The sandbox example will be pushed after the PR merge because it depends on the image `envoyproxy/envoy-contrib-dev:latest` which contains the go extension feature.

Reply:
Resolved it. Thanks.